src/mongoc/mongoc-topology-scanner.c

```diff
@@ -0,0 +1,396 @@
+
+static void
+mongoc_topology_scanner_node_destroy (mongoc_topology_scanner_node_t *node);
+
+mongoc_topology_scanner_t *
+mongoc_topology_scanner_new (mongoc_topology_scanner_cb_t cb,
+                             void                        *cb_data)
+{
+   mongoc_topology_scanner_t *ts = bson_malloc0 (sizeof (*ts));
+
+   ts->async = mongoc_async_new ();
+   bson_init (&ts->ismaster_cmd);
+   BSON_APPEND_INT32 (&ts->ismaster_cmd, "isMaster", 1);
+
+   ts->cb = cb;
+   ts->cb_data = cb_data;
+
+   return ts;
+}
+
+void
+mongoc_topology_scanner_destroy (mongoc_topology_scanner_t *ts)
+{
+   mongoc_topology_scanner_node_t *ele, *tmp;
+
+   DL_FOREACH_SAFE (ts->nodes, ele, tmp) {
+      mongoc_topology_scanner_node_destroy (ele);
+   }
+
+   mongoc_async_destroy (ts->async);
+   bson_destroy (&ts->ismaster_cmd);
+
+   bson_free (ts);
+}
+
+void
+mongoc_topology_scanner_add (mongoc_topology_scanner_t *ts,
+                             const mongoc_host_list_t  *host,
+                             uint32_t                   id)
+{
+   mongoc_topology_scanner_node_t *node = bson_malloc0 (sizeof (*node));
+
+   memcpy (&node->host, host, sizeof (*host));
+
+   node->id = id;
+   node->ts = ts;
+
+   DL_APPEND(ts->nodes, node);
+
+   return;
+}
+
+static void
+mongoc_topology_scanner_node_destroy (mongoc_topology_scanner_node_t *node)
+{
+   DL_DELETE (node->ts->nodes, node);
+
+   if (node->dns_results) {
+      freeaddrinfo (node->dns_results);
+      node->dns_results = NULL;
+      node->current_dns_result = NULL;
+   }
+
+   if (node->cmd) {
+      mongoc_async_cmd_destroy (node->cmd);
+   }
+
+   if (node->stream) {
+      mongoc_stream_destroy (node->stream);
+   }
+
+   bson_free (node);
+}
+
+void
+mongoc_topology_scanner_rm (mongoc_topology_scanner_t *ts,
+                            uint32_t                   id)
+{
+   mongoc_topology_scanner_node_t *ele, *tmp;
+
+   DL_FOREACH_SAFE (ts->nodes, ele, tmp)
+   {
+      if (ele->id == id) {
+         mongoc_topology_scanner_node_destroy (ele);
+         break;
+      }
+
+      if (ele->id > id) {
+         break;
+      }
+   }
+}
+
+/*
+ *-----------------------------------------------------------------------
+ *
+ * This is the callback passed to async_cmd when we're running
+ * ismasters from within the topology monitor.
+ *
+ *-----------------------------------------------------------------------
+ */
+
+static void
+mongoc_topology_scanner_ismaster_handler (mongoc_async_cmd_result_t async_status,
+                                          const bson_t             *ismaster_response,
+                                          int64_t                   rtt_msec,
+                                          void                     *data,
+                                          bson_error_t             *error)
+{
+   mongoc_topology_scanner_node_t *node;
+
+   bson_return_if_fail (data);
+
+   node = (mongoc_topology_scanner_node_t *)data;
+   node->cmd = NULL;
+
```

Comment:
we should explicitly handle a MONGOC_ASYNC_CMD_TIMEOUT here, even though we currently pass a NULL ismaster_response in that case.